### PR TITLE
Pipeline optimization

### LINKED ISSUE #80: Pipeline optimization
....

### PR DESCRIPTION
**Pull request by jcorreia11**
# DeepMol Pipeline Optimization

In DeepMol we can optimize Pipelines using the `PipelineOptimization` class. This class
uses the `Optuna` library to optimize the hyperparameters of the pipeline. It is possible
to optimize between different `Transformers` (e.g. featurizers, scalers, etc.) and 
`Predictors` (i.e. models) and also to optimize their hyperparameters.

For that we need to define an objective function that will be optimized by `Optuna`. This
objective function will receive a `Trial` object from `Optuna` and will return the
steps (i.e. `Transformers` and `Predictors`) of the pipeline. Alternatively, we can
use a set of predefined objective functions (presets) that are already implemented in
DeepMol. The following presets are available:
- `'sklearn'`: optimizes between all available Standardizers, Featurizers, Scalers,
            Feature Selectors in DeepMol and all available models in Sklearn.
- `'deepchem'`: optimizes between all available Standardizers and all available models
              in DeepChem (depending on the model different Featurizers, Scalers 
              and Feature Selectors may also be optimized).
- `'keras'`: optimizes between Standardizers, Featurizers, Scalers, Feature Selectors in 
           DeepMol and some base Keras models (1D CNNs, RNNs, Bidirectional RNNs and FCNNs).
- `'all'`: optimizes between all the above presets.

In the case of using a custom objective function, the function must return a list of tuples
where the first element of the tuple is the name of the step and the second element is the
object that implements the step. The pipeline class will respect the order of the steps you
return in the objective function. For example, if you return the following list of tuples:

```python
steps = [('featurizer', featurizer), ('scaler', scaler), ('model', model)]
```

The pipeline will first apply the featurizer, then the scaler and finally the model.

## Example using a custom objective function:

In the following example, we will assume that you already have the data processed (ready for
training). We will only use DeepMol's PipelineOptimization class to optimize the final
model. In this case, we will optimize between a Random Forest and a SVC model. We will also
optimize the hyperparameters of the models.

```python
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC

from deepmol.loaders import CSVLoader
from deepmol.metrics import Metric
from deepmol.models import SklearnModel
from deepmol.pipeline_optimization import PipelineOptimization
from deepmol.splitters import RandomSplitter

# DEFINE THE OBJECTIVE FUNCTION
def objective(trial):
    model = trial.suggest_categorical('model', ['RandomForestClassifier', 'SVC'])
    if model == 'RandomForestClassifier':
        n_estimators = trial.suggest_int('model__n_estimators', 10, 100, step=10)
        model = RandomForestClassifier(n_estimators=n_estimators)
    elif model == 'SVC':
        kernel = trial.suggest_categorical('model__kernel', ['linear', 'poly', 'rbf', 'sigmoid'])
        model = SVC(kernel=kernel)
    model = SklearnModel(model=model, model_dir='model')
    steps = [('model', model)]
    return steps
 
# LOAD THE DATA   
loader = CSVLoader('data_path...',
                   smiles_field='mols',
                   id_field='ids',
                   labels_fields=['y'],
                   features_fields=['f1', 'f2', 'f3', '...'],
                   mode='classification')
dataset_descriptors = loader.create_dataset(sep=",")
   
# OPTIMIZE THE PIPELINE 
po = PipelineOptimization(direction='maximize', study_name='test_predictor_pipeline')
metric = Metric(accuracy_score)
train, test = RandomSplitter().train_test_split(dataset_descriptors, seed=123)
po.optimize(train_dataset=train, test_dataset=test, objective_steps=objective, 
            metric=metric, n_trials=5, save_top_n=3)
``` 

This will optimize the objective function and will save the top 3 pipelines
(save_top_n=3) in the `study_name` folder. The `direction` parameter indicates if we want
to maximize or minimize the metric. In this case, we want to maximize the accuracy. The
`n_trials` parameter indicates the number of trials that will be performed by `Optuna`.

Additionally, we can also provide storage (see: https://optuna.readthedocs.io/en/stable/reference/storages.html),
a sampler (https://optuna.readthedocs.io/en/stable/reference/samplers/index.html) and a
pruner (https://optuna.readthedocs.io/en/stable/reference/pruners.html) to use in the
optimization.

The best parameters can be retrieved using the `best_params` property:

```python
best_params = po.best_params
```

The best trial can be retrieved using the `best_trial` property:

```python
best_trial = po.best_trial
```

The best score can be retrieved using the `best_value` property:

```python
best_score = po.best_value
```

All the trials can be retrieved using the `trials` property:

```python
trials = po.trials
```

The best pipeline can be retrieved using the `best_pipeline` property:

```python
best_pipeline = po.best_pipeline
```

More information about the trials and respective scores and parameters can be retrieved
using the `trials_dataframe` method.

```python
trials_df = po.trials_dataframe()
```

Information about the importance of the optimized parameters can be retrieved using the
`get_param_importances` method.

```python
param_importances = po.get_param_importances()
```

## Example using a preset:

The following example assumes that you are providing raw SMILES data and respective labels.
We will use the `'all'` preset.

```python
# LOAD THE DATA
loader = CSVLoader('dataset_regression_path',
                   smiles_field='smiles',
                   labels_fields=['pIC50'],
                   mode='regression')
dataset_regression = loader.create_dataset(sep=",")

# OPTIMIZE THE PIPELINE
po = PipelineOptimization(direction='minimize', study_name='test_pipeline')
metric = Metric(mean_squared_error)
train, test = RandomSplitter().train_test_split(dataset_regression, seed=123)
po.optimize(train_dataset=train, test_dataset=test, objective_steps='all', 
            metric=metric, n_trials=10, data=train, save_top_n=2)
```

In this case, we are optimizing between all the available steps in DeepMol in order to minimize the mean squared error (regression task).

The best params and best pipeline can be obtained as in the previous example.

Closes #80 .

### CODE REVIEW

**Review by jcapels — APPROVED** [review]
I will approve this PR. I am impressed! Congratulations on the work.

I have just a few questions and a few picky things about the code. 
I think we should put the TODOs of the **src/deepmol/pipeline_optimization/_utils.py** in the issues. 
Although impressively you were able to put together so many models, it can create a problem: it can force us to validate everything with the case studies and create comprehensive documentation.  We should plan if we do this now, after the deadline of the first paper or not do it at all and only do it for the most relevant ones.

File: src/deepmol/models/keras_model_builders.py [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+    mode = 'classification' if 'mode' not in keras_kwargs else keras_kwargs['mode']
+    epochs = 150 if 'epochs' not in keras_kwargs else keras_kwargs['epochs']
+    batch_size = 10 if 'batch_size' not in keras_kwargs else keras_kwargs['batch_size']
+    verbose = 0 if 'verbose' not in keras_kwargs else keras_kwargs['verbose']
+    model_kwargs = {} if model_kwargs is None else model_kwargs
+    return KerasModel(model_builder=baseline_dense_model_builder, mode=mode, model_dir=model_dir, epochs=epochs,
+                      batch_size=batch_size, verbose=verbose, **model_kwargs)
+
+
+def baseline_cnn_model_builder(input_dim: int = 1024,
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/models/keras_model_builders.py:103]: done

File: src/deepmol/models/keras_model_builders.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+        model.add(Conv1D(filters[i], kernel_sizes[i], activation=activations[i], strides=strides[i], padding=padding))
+    model.add(Flatten())
+    model.add(Dropout(dropout))
+    model.add(Dense(dense_units, activation=dense_activation))
+    model.add(Dense(last_layer_units, activation=last_layer_activation))
+    model.compile(loss=loss, optimizer=optimizer, metrics=metrics)
+    return model
+
+
+def keras_cnn_model(model_dir: str = 'keras_model/', model_kwargs: dict = None,
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/models/keras_model_builders.py:174]: done

File: src/deepmol/pipeline_optimization/_featurizer_objectives.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+from typing import Literal
+
+from optuna import Trial
+
+from deepmol.base import Transformer
+from deepmol.compound_featurization import All3DDescriptors, TwoDimensionDescriptors, MorganFingerprint, \
+    AtomPairFingerprint, LayeredFingerprint, RDKFingerprint, MACCSkeysFingerprint, Mol2Vec, SmilesOneHotEncoder
+
+# TODO: include All3DDescriptors? (It takes a lot of time to compute)
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_featurizer_objectives.py:9]: yeah, I think it is best.

File: src/deepmol/pipeline_optimization/_featurizer_objectives.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# TODO: include MixedFeaturizer?
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_featurizer_objectives.py:10]: I added the mixed featurizer with 2 featurizers that can be optimized with optuna

File: src/deepmol/pipeline_optimization/_sklearn_model_objectives.py [hunk 5]
```diff
@@ -0,0 +1,1863 @@
+from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier, GradientBoostingRegressor, \
+    GradientBoostingClassifier
+from sklearn.linear_model import LinearRegression, Ridge, Lasso, LogisticRegression
+from sklearn.neural_network import MLPRegressor, MLPClassifier
+from sklearn.svm import SVR, SVC
+
+from deepmol.base import Predictor
+from deepmol.models.sklearn_model_builders import linear_regression_model, ridge_model, ridge_classifier_model, \
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_sklearn_model_objectives.py:8]: done

File: src/deepmol/pipeline_optimization/_utils.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
+from copy import deepcopy, copy
+from typing import Literal
+
+from deepchem.models import TextCNNModel
+
+from deepmol.base import DatasetTransformer, PassThroughTransformer
+from deepmol.compound_featurization import SmilesSeqFeat, CoulombFeat
+from deepmol.datasets import Dataset
+from deepmol.pipeline_optimization._deepchem_models_objectives import gat_model_steps, gcn_model_steps, \
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_utils.py:9]: done

File: src/deepmol/pipeline_optimization/_utils.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    data : deepmol.datasets.Dataset
+        The dataset.
+
+    Returns
+    -------
+    list
+        The list of steps for the deepchem preset.
+    """
+    # TODO: "mpnn_model" is not working (DeepChem-> AttributeError: 'GraphData' object has no attribute 'get_num_atoms')
+    # TODO: "megnet_model" is not working (error with torch_geometric (extra_requirement))
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_utils.py:66]: yes, forgot to remove todo

[the remark this answers is not in the document]
Comment (jcorreia11) [reply src/deepmol/pipeline_optimization/_utils.py:66]: oh, i am recalling now that it returned some error related to the torch_geometric (maybe later try with different versions)

File: docs/deepmol_docs/pipeline_optimization.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
+****# DeepMol Pipeline Optimization
```
[the remark this answers is not in the document]
Comment (jcorreia11) [reply docs/deepmol_docs/pipeline_optimization.md:1]: i think this was added automatically for some reason, but i will remove it :D

**Pull request merged**